parlai/agents/transformer/module_extensions.py

```diff
@@ -0,0 +1,195 @@
+import torch.nn.functional as F
+from contextlib import AbstractContextManager
+
+from parlai.core.params import ParlaiParser
+
+R3F_NOISE_NORMAL = "normal"
+R3F_NOISE_UNIFORM = "uniform"
+
+
+class R3FNoiseContext(object):
+    """
+    Class that helps implement "Better Fine-Tuning by Reducing Representational
+    Collapse", Aghajanyan et al. (2020). https://arxiv.org/ans/2008.03156.
+
+    This class should be instantiated at a point that has access to command line args
+    and the `compute_loss` function.
+    """
+
+    @staticmethod
+    def add_cmdline_args(argparser: ParlaiParser):
+        group = argparser.add_argument_group('R3F fine-tuning Args')
+        group.add_argument(
+            '--use-r3f',
+            type=bool,
+            default=False,
+            help='should we use the R3f loss at all?',
+        )
+        group.add_argument('--r3f-eps', type=float, default=1e-5, help='noise eps')
+        group.add_argument(
+            '--r3f-lambda',
+            type=float,
+            default=1.0,
+            help='lambda for combining logistic loss and noisy KL loss',
+        )
+        group.add_argument(
+            '--r3f-noise-type',
+            type=str,
+            default=R3F_NOISE_UNIFORM,
+            choices=[R3F_NOISE_NORMAL, R3F_NOISE_UNIFORM],
+            help='type of noises for RXF methods',
+        )
+        group.add_argument(
+            '--r3f-encoder-noise',
+            type=bool,
+            default=True,
+            help='Add noise to encoder. At least one of `r3f-encoder-noise` and `r3f-coder-noise` must be set to True',
+        )
+        group.add_argument(
+            '--r3f-decoder-noise',
+            type=bool,
+            default=False,
+            help='Add noise to decoder. At least one of `r3f-encoder-noise` and `r3f-coder-noise` must be set to True',
+        )
+
+    def __init__(self, opts):
+        self.use_r3f = opts.get('use_r3f')
+        self.eps = opts.get('r3f_eps')
+        self.noise_type = opts.get('r3f_noise_type')
+        if self.noise_type == R3F_NOISE_NORMAL:
+            self.noise_sampler = torch.distributions.normal.Normal(
+                loc=0.0, scale=self.eps
+            )
+        elif self.noise_type == R3F_NOISE_UNIFORM:
+            self.noise_sampler = torch.distributions.uniform.Uniform(
+                low=-self.eps, high=self.eps
+            )
+        self.r3f_lambda = opts.get("r3f_lambda")
+        self.noise_encoder = opts.get("r3f_encoder_noise")
+        self.noise_decoder = opts.get("r3f_decoder_noise")
+
+        if self.use_r3f and not self.noise_encoder and not self.noise_decoder:
+            raise RuntimeError(
+                "R3FContext: Noise must be added to at least one of the encoder or decoder."
+            )
+        self.had_first_pass = (
+            False
+        )  # necessary because nn.Embedding for encoder/decoder on some models are the same. Use this to only make any relevant deep copies once.
```

Comment:
I'm hoping we can reach a conclusion on which mode is best in your experiments (encoder only, decoder only, both), and then make an opinionated implementation which does only that.

Reply:
Makes sense. 

This would still be necessary in the encoder only case (unless doing a deep copy of an embedding each train loop isn't too bad?)

Reply:
Makes sense.

Reply:
I think a deep copy is very expensive

If we end up going encoder-only, the hook can set the first pass to True to turn any future calls to the hook into the identity function

Reply:
Yeah, makes sense as a workaround.